### PR TITLE
Internal improvement: Fully convert compile-solidity to use TypeScript

### PR DESCRIPTION
**Pull request by eggplantzzz**
NOTE THAT THIS IS A BREAKING CHANGE FOR @truffle/compile-common.

Note that types were moved from `compile-common/src/types.ts` to `compile-common/src/sources.ts` in order to prevent a clobbering issue. `src/index.ts` is the following:
```
export * as Shims from "./shims";
export * as Sources from "./sources";
export * as Errors from "./errors";
export * as Compilations from "./compilations";
export * from "./types";
```

If the type `Sources` is exported from `./types` then it will conflict with the `Sources` namespace.

### CODE REVIEW

File: packages/compile-common/src/types.ts
Comment (gnidan) [inline packages/compile-common/src/types.ts:56]: This makes this PR a breaking change to compile-common, yes?

Comment (eggplantzzz) [reply packages/compile-common/src/types.ts:56]: Will this be breaking? It makes the type more lenient meaning that it shouldn't break anything, right?

Comment (benjamincburns) [reply packages/compile-common/src/types.ts:56]: Yeah, should be considered non-breaking, IMO

Comment (gnidan) [reply packages/compile-common/src/types.ts:56]: "Making the type more lenient" only applies if this optionality is an input type. Like, if we're talking about exporting some function`(options: { /* ... */ }) => any`, then adding a `?` to one of the `options` is non-breaking.

Here, it's different... to a package consuming the type, upgrading to the version that includes this change means handling a new case: previously, (e.g.) you would only need to handle `object`, but now you must handle both sides of `object | undefined`.

File: packages/compile-common/src/sources.ts
Comment (benjamincburns) [inline packages/compile-common/src/sources.ts:17]: Why remove these from the `types` file?

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/profiler/loadParser.ts:39]: Is the return type of `semver.coerce` strictly `Semver | null`? If not, you likely either want `Optional<SemVer>` or `SemVer?` here.

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/profiler/loadParser.ts:56]: Is there not a better type for this? If not, maybe this should be a generic?

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/profiler/loadParser.ts:66]: Logic change. Are we sure that we want `parser.version` instead of `solc.version` here? Can they ever be different versions?

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/profiler/loadParser.ts:56]: If there really is no better type for this apart from `any`, or if defining a valid type constraint as part of this PR would be high effort, you might also consider defining a `SolidityParser` type alias that is still `any`, as that way it'll be easy to update in the future should better types become available.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:148]: Why remove the typedoc?

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:69]: Minor nit - you might consider moving these to a local `types.ts` file, which would be in line with other TS packages in the repo.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:81]: If we are okay with breaking changes here, you _might_ want to consider making this throw instead of returning `null`, depending on what else calls this. That would be my recommendation, unless we do want this to be a strictly homomorphic change.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:148]: Ah, I guess they weren't all that useful before. Might be nice if you were to update them, though.

That said, consider this an optional request.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:237]: You might want to define `XOptions` interfaces for these args, as personally I find this syntax to be far less readable. There are a few other occurrences of this, but I'll just flag this one to cut down on the noise.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:285]: How does this compile without a type definition for the args? I think you may have a bunch of things that are implicitly typed as `any`.

I'd recommend turning on the `noImplicitAny` flag in `tsconfig.json` and seeing what pops up.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:351]: I don't think the use of the `!` operator is safe here. I'd recommend using a type assertion instead.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:507]: this would be a good place to use the coalescing operator
```suggestion
            immutableReferences: deployedBytecodeInfo?.immutableReferences,
```

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:526]:
```suggestion
    .mapFlat(source => Object.keys(source));
```

File: packages/core/lib/testing/Test.js
Comment (benjamincburns) [inline packages/core/lib/testing/Test.js:16]: This is an example of a change that wouldn't be necessary if you were to use the `export =` syntax in `compile-solidity/dist/profiler`

File: yarn.lock
Comment (benjamincburns) [inline yarn.lock:6529]: BTW it's always worth reading your `yarn.lock` to see what actually changed.

If you see a new line for a module that's already in use elsewhere in the repo, that's a _strong_ indication that you're introducing/increasing dependency version inconsistency.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:34]: Can this be given a more specific type?

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:43]: Can these be given a more specific type?

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:49]: Can this be given a more specific type?

File: packages/compile-solidity/src/compileWithPragmaAnalysis.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/compileWithPragmaAnalysis.ts:149]: I think it can be typed as `Compilation[]`, given the return type of the `run` function that populates this array.

**Review by benjamincburns — CHANGES_REQUESTED** [review]
Nice work overall - I _really_ love seeing JS code swapped over.

I understand that some of the `any` types that I flagged might be challenging, so please do feel free to drop a note on any of them that are too challenging to fix now. I think my only ask would be that where that's the case, we should raise issues so that we can improve the types here over time.

Otherwise I think the only thing that I didn't address in my comments is that I think it'd be very nice if the tests were also converted over to TS. Totally not necessary as part of this PR, but I just wanted to give a nudge all the same.

File: packages/compile-common/src/types.ts
Comment (benjamincburns) [reply packages/compile-common/src/types.ts:56]: Oh, very nice catch! I should've realized that this is a return type.

File: packages/compile-solidity/src/index.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/index.ts:211]: On further review, I don't think that this change in export syntax _does_ introduce a breaking change _here_. As such, it might be a better idea to delete the explicit `export` line and switch to using the `export` keyword inline with the definitions.

Please do take this as a nudge to review the export syntax in other modules (such as in `compile-solidity/dist/profiler`)

**Comment by benjamincburns** [comment]
Oh one other thing - I'd recommend squashing these commits immediately prior to merging (but after review approval). I see at least one `update yarn.lock` commit, which leads me to believe that there are probably commits added by this PR that don't build properly, and that makes tools like `git bisect` a lot harder to use.

File: packages/compile-common/src/sources.ts
Comment (eggplantzzz) [reply packages/compile-common/src/sources.ts:17]: `Sources` was being clobbered in that other location due to this namespace being named `Sources`. So in other words, these are the exports for everything in compile-common/src...
```
export * as Shims from "./shims";
export * as Sources from "./sources";
export * as Errors from "./errors";
export * as Compilations from "./compilations";
export * from "./types";
```

The `Sources` type was being clobbered by the `Sources` namespace. @gnidan advised me that this is a perfectly good new location.

File: packages/compile-solidity/src/compileWithPragmaAnalysis.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/compileWithPragmaAnalysis.ts:158]: This along with the other instances were necessary to handle `compilation` being `null`.

File: packages/compile-solidity/src/index.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/index.ts:43]: So I'm not sure we want to throw in this instance. The `run` method returns a `Compilation` or `null` if there is nothing to compile. So if there was nothing to compile in `run`, `null` is returned and we don't push the result into the array.

File: packages/compile-solidity/src/index.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/index.ts:81]: This seems like the same to me, TS didn't like doing it the other way but I forget why.

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/profiler/loadParser.ts:39]: Yup, it is strictly `Semver | null`. https://github.com/npm/node-semver/blob/main/functions/coerce.js

File: packages/compile-solidity/src/index.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/index.ts:43]: So I'll just respond to all the comments of this type here. I think the old behavior is buggy. Should it return `null` it would throw in the old code as you can't access `contracts` on `null`. So I was considering this fixing a bug.

File: packages/compile-solidity/src/index.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/index.ts:211]: okie doke

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/profiler/loadParser.ts:66]: This was an error. There is no `solc` in this scope and rather `parser` is the variable we want to use. If you look at the call to this method ~ line 24 you can see `solc` is passed in here (a portion of the result of calling `supplier.load`). Looks to me like someone got their names confused.

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts:148]: It seemed redundant now that we have the types...plus TS told me to get rid of it :)

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts:525]: That would be nice but I didn't implement that because `flatMap` isn't supported in Node 12. I assume that ends up in the built code.

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts:351]: So in a lot of these situations I just tried to maintain current behavior. You think I should make it throw? I suppose that would be an improvement.

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts:237]: sure, I don't mind

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts]: `Compile.Sources.PathMapping`

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts]: I'm not certain about this as this is something returned from `solcWrap`. If you follow it back, it returns whatever gets returned [here](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts#L119-L129). I don't know what that is exactly, any suggestions?

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts]: I changed these to `object` for now as I'm not certain how accurate I can be about what will be in here to be honest.

File: packages/compile-solidity/src/index.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/index.ts:44]: Yeah if you're confident that everything works fine this way, then it's fine by me.

File: packages/compile-solidity/src/index.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/index.ts:82]: Ah my bad, for some reason I didn't catch that these were meant to be one-and-only pushes to the array.

I do find the old syntax to be more clear as to that intention, though.

If TS complained, it's probably something that the `as` keyword could fix, but I'm not super fussed about this now that I understand that this is logically correct.

File: packages/compile-common/src/sources.ts
Comment (benjamincburns) [reply packages/compile-common/src/sources.ts:17]: Makes sense. Might be worth adding a comment, but I don't think it's strictly necessary.

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/profiler/loadParser.ts:66]: That begs the question as to whether this code is even reachable... Not something to worry about in this PR if we're aiming for a homomorphic change here, however.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:148]: Yeah if there isn't more to say other than type details, I suppose it would be.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:36]: Not important to me. I wanted to be sure that you'd considered it, but you _definitely_ have now, and it does seem reasonable to leave it as is.

File: packages/compile-common/src/sources.ts
Comment (eggplantzzz) [reply packages/compile-common/src/sources.ts:17]: Probably the only thing useful in the future will be the conversation on this PR. Nobody is going to know that these types used to be in `./types.ts`, right?

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:476]: Pretty sure it down-compiles to a call to `map(...).flat()` thanks to the target being `ES2016`, but to be sure you'd have to check the JS output. That said, it's not really worth the extra time.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts]: I'm not too fussed.

Worst case you could define a custom interface here that at least captures the members of the type on which you depend, but I don't see a ton of value in that, as it would make more sense for there to be actual types coming from `solcWrap`.

I don't think there is a good answer to this, but the only lingering concern that I have is over what would happen if there was suddenly proper typing for `solcWrap`? Ideally this dependency would be immediately discoverable due to a compiler error, but that won't happen if it's `any`.

I suppose a custom local type would break if it had structural conflicts with the hypothetical typed return value of `solcWrap`, but I dunno - I don't like this package assuming authority over another package's types.

So yeah, this is probably just a really long-winded way of saying "yeah, fair enough." 😆

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts]: If this code doesn't depend on any properties of these (sorry, I didn't go look again to see if it does), then that's maybe fine? Alternatively you could consider typing anything that is opaque to this package as `unknown`, as that would force a consumer to declare a type for it before actually consuming it.

I'll leave the decision as to whether or not that is appropriate to you, however.

File: packages/compile-common/src/sources.ts
Comment (benjamincburns) [reply packages/compile-common/src/sources.ts:17]: Yeah fair enough.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:302]: Yeah, it throws either way if the `!` is invalid. If you throw on a null assertion, at least you can provide some context.

Otherwise if we leave it strictly identical in behavior in cases like this, we're negating the primary value that the TS type checker provides, IMO.

Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:302]: Also an explicit throw is more readable, IMO, as that makes it a lot more clear that this code _can_ throw.

File: packages/core/lib/testing/Test.js
Comment (eggplantzzz) [reply packages/core/lib/testing/Test.js:16]: I like named imports :)

**Comment by eggplantzzz** [comment]
Thanks for the review @benjamincburns! Oh, and as for converting the tests to TS, see https://github.com/trufflesuite/truffle/pull/5445 where the tests get removed from this package altogether :)

File: packages/compile-solidity/src/profiler/loadParser.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/profiler/loadParser.ts:66]: Do you think I'm mistaken? I think this code will crash if executed. I would argue that this is a small enough change that it shouldn't be a big deal to include it. What do you think about that point of view?

Comment (benjamincburns) [reply packages/compile-solidity/src/profiler/loadParser.ts:66]: > Do you think I'm mistaken?

Oh not at all! My reply was more academic, sorry. It's more that if the original code would crash if executed, but we weren't ever seeing that, this whole path may be effectively dead code.

I don't think there's any further action required here. Sorry for the confusion!

File: packages/core/lib/testing/Test.js
Comment (benjamincburns) [reply packages/core/lib/testing/Test.js:16]: Yeah totally fair since this is an internal module. Just highlighted it more as an example of the difference caused by the way that TS handles exports. Apologies if this was already well understood on your end. I only recently encountered this "restriction" of the TS export syntax.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:236]: FWIW, since this code bumps up against a bunch of stuff that's untyped, please don't interpret my suggestion here as trying to require you to enable `noExplicitAny` in order to merge this. I do think this code would probably be improved if `noImplicitAny` were `true`, but I was more just suggesting temporarily enabling it to spot any implicit `any` types to make it easier to decide whether it's possible/appropriate to make them explicitly typed.

Net net, I think this PR is already well over the threshold of value added to our code, so I'm likely to approve once you're done pushing changes.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:103]: Sorry that I missed this on my first pass, but I think this change is breaking? That is, assuming that we consider the `run` function part of the public interface of the package.

I'd say it's breaking because code that previously expected an `undefined` for `sourceIndexes` in the return data must now be changed to expect an empty array.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [inline packages/compile-solidity/src/run.ts:148]: I don't think `object` is an appropriate type here, is it? That is, shouldn't it be an object with the same index signature that was defined by the map?

```suggestion
    .reduce((a: { [functionName: string]: number } , b: { [functionName: string]: number }) => Object.assign({}, a, b), {});
```

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts:103]: I would say that it probably isn't part of the interface. It is not exposed by it. It is exposed, however, in the work done to move the tests [here](https://github.com/trufflesuite/truffle/pull/5445). But that is purely so that it can be tested by this separate packages. What do you think @gnidan? Maybe you have a different opinion.

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts:148]: Ok, that is a bit more precise.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:103]: Assuming that the consensus is that these _aren't_ part of the public interface, but that they still need to be exported for testing purposes, I think we should consider using some naming prefix to indicate to would-be consumers that these are not actually part of the public interface.

Historically I've used an underscore or dunder prefix for that, but I'm not really all that fussed about what we actually use - so long as it's communicated externally.

Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:103]: Alternatively maybe we just communicate that unless otherwise documented, only the things exported by the module's entrypoint should be considered part of the public interface?

File: packages/compile-solidity/src/run.ts
Comment (eggplantzzz) [reply packages/compile-solidity/src/run.ts]: Do we want to force the user to do that?

File: packages/compile-solidity/src/run.ts
Comment (gnidan) [reply packages/compile-solidity/src/run.ts:103]: > Alternatively maybe we just communicate that unless otherwise documented, only the things exported by the module's entrypoint should be considered part of the public interface? This is probably much easier to enforce in practice.

This is my strong preference - don't force us to use underscores ;)

Particularly following @wbt's excellent concerns raised in https://github.com/trufflesuite/truffle/discussions/5417 (aside: sorry wbt; I know I still owe you a response there), I think we should probably write something similar to how Hardhat has its ["stability guarantees" page](https://hardhat.org/hardhat-runner/docs/reference/stability-guarantees), where it says at the top that they don't "follow semver strictly."

This will probably hurt a bit at first, but it seems best to maintain our public interfaces explicitly as "what is documented is what you can rely on". Sure, sometimes people like you, @wbt, will have to fix breakages as they happen, but that's unavoidable... the only user-facing difference with strict adherence to semver would be that each breakage gets a major release; the work to upgrade remains the same. But it's costly to do major releases, and it's also dangerous for us to let our public interfaces remain so loosely defined as "whatever is exported"... maybe in a few years we'll have the project maturity to support that, but right now this sort of thing keeps causing us to get in our own way.

Comment (gnidan) [reply packages/compile-solidity/src/run.ts:103]: (Oh and please don't fret @wbt; I'm not talking about your recently raised issues about our breaking types. Those are absolutely mea culpa and should've been major releases or not at all. I'm specifically talking here about the part of the interface surface area that we believe is internal)

Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:103]: > This is my strong preference - don't force us to use underscores ;)

Ha, I'm not sure that I could even if I wanted to!

Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts:103]: I'm going to resolve this thread under the assumption that this isn't a change to public interface, and that @gnidan and/or @eggplantzzz will handle figuring out how to communicate our stability guarantees elsewhere.

File: packages/compile-solidity/src/run.ts
Comment (benjamincburns) [reply packages/compile-solidity/src/run.ts]: I'd vote yes. Using `unknown` for code that hasn't been typed yet creates a sort of boundary line for unchecked types in our code. If we use `any`, the unchecked types propagate out far-and-wide, and it can become very difficult to reign it in later.

That said, if there's a difference of opinion here, I'm not staunch enough about this to block the entire changeover to TS just for this one concern.

Here's some reading on the subject, should you want to consider it further.
https://dmitripavlutin.com/typescript-unknown-vs-any/

**Review by benjamincburns — APPROVED** [review]
Approved on the assumption that there's consensus that the `run` function is internal code. If this isn't a correct assumption, please don't merge until that's resolved.

I also left a few comment threads unresolved, but I don't think that resolving them is necessary to get this merged.

Super happy to be seeing more and more of our code converting over to TS!

**Comment by wbt** [comment]
I'm not sure why I got a mention notification on merge, but please consider this a polite reminder that breaking changes should be marked as such in the [semver](https://github.com/orgs/trufflesuite/discussions/5417) update!

**Pull request merged**
